agreementbot/object_policy.go

```diff
@@ -0,0 +1,248 @@
+		if err != nil {
+			glog.Errorf(opLogstring(fmt.Sprintf("unable to read agreements, error %v", err)))
+		}
+
+		for _, agreement := range agreements {
+			// If an existing policy has changed, and the agreement's node is a destination in the old policy, then check to see if the node
+			// is still compatible with the new policy. If not, remove the node from the object's destination list.
+			if cmd.Msg.OldPolicy != nil &&  cutil.SliceContains(oldPolicyDestNodes, agreement.DeviceId) {
+
+				// Convert the object's policy into an internal policy so that we can do the compatibility check.
+				internalObjPol := policy.Policy_Factory(fmt.Sprintf("object policy for %v type %v", newPolicy.ObjectID, newPolicy.ObjectType))
+				internalObjPol.Properties = newPolicy.DestinationPolicy.Properties
+				internalObjPol.Constraints = newPolicy.DestinationPolicy.Constraints
+				glog.V(5).Infof(opLogstring(fmt.Sprintf("converted new object policy to: %v", internalObjPol)))
+
+				nodePolicy,err := w.GetNodePolicy(agreement.DeviceId)
+				if err != nil {
+					glog.Errorf(opLogstring(fmt.Sprintf("%v", err)))
+				} else if err := policy.Are_Compatible(nodePolicy, internalObjPol); err != nil {
+					// This agreement's node is no longer compatible, remove it from the destination list of the object.
+					if err := UnassignObjectFromNode(w, &newPolicy, agreement.DeviceId); err != nil {
+						glog.Errorf(opLogstring(fmt.Sprintf("%v", err)))
+					}
+				} else {
+					glog.V(5).Infof(opLogstring(fmt.Sprintf("node %v is still compatible with object %v with type %v", agreement.DeviceId, newPolicy.ObjectID, newPolicy.ObjectType)))
+				}
+			} else {
+
+			// 1. Brand new policy - node might be eligble now
+			// 2. Updated policy - node might be eligible now
+
+
+			// If the agreement has a service id in the list of services for the newpolicy AND the agreement's node is not in the dest list of the old policy,
+			// then check to see if the new policy is compatible with node policy. If so, add the object to the node.
+				objPolicies := new(exchange.ObjectDestinationPolicies)
+				(*objPolicies) = append((*objPolicies), newPolicy)
+				for _, serviceId := range agreement.ServiceId {
+
+					serviceNamePieces := strings.SplitN(serviceId, "_", 2)
+					if cutil.SliceContains(newPolicyServiceKeys, serviceNamePieces[0]) && !cutil.SliceContains(oldPolicyDestNodes, agreement.DeviceId) {
+						// Add the node to the object destination if eligible.
+						nodePolicy,err := w.GetNodePolicy(agreement.DeviceId)
+						if err != nil {
+							glog.Errorf(opLogstring(fmt.Sprintf("%v", err)))
+						} else if err := AssignObjectToNode(w, objPolicies, agreement.DeviceId, nodePolicy); err != nil {
+							glog.Errorf(opLogstring(fmt.Sprintf("%v", err)))
+						}
+						break
+					}
+				}
+			}
+
+		}
+	}
+
+	return
+
+}
+
+// Get node policy
+func (w *AgreementBotWorker) GetNodePolicy(deviceId string) (*policy.Policy, error) {
```

Comment:
There is a same function in BaseAgreementWorker.  Should we create one without class name instead?

Reply:
Yes, that's exactly what I will do in my next PR. There is a bug in the sync service code that is preventing me from finishing that section of event handling code. When that bug is resolved, I will be able to know if I still need that function to be used by both the main worker and the protocol workers.